### ISSUE TITLE
Add ChaCha20 Support

### ISSUE THREAD

**Issue by librehat** [shadowsocks/shadowsocks-chromeapp#14]
[chacha20.js](https://gist.github.com/devi/da696f47865605e5f6ed) can be helpful.

I may contribute to this project once my laptop arrives

**Comment by meowlab** [comment]
Thanks~
I'll take a look later and PR is welcomed~

**Comment by Jason110go** [comment]
Hey，how's it going? Sounds like you really made a great move!

**Comment by nekolab** [comment]
Sorry, currently I have no time to work on this : (
Since I'm planning to use native client engine replace js engine, so this issue may have lower priority
As always, if you want to give some help, pull request is welcome

**Comment by librehat** [comment]
I tried to use [`chacha20.js`](https://github.com/quartzjer/chacha20) but couldn't make it work. ;( You might wish to check my WIP commit: https://github.com/librehat/shadowsocks-chromeapp/commit/e53c3015edf933e574d4766cd6c8acd1cfef3924

**Comment by nekolab** [comment]
@librehat Maybe we need to set a padding to create a block and set a counter like the implementation in C version?

**Comment by librehat** [comment]
@meowlab I don't think we need to set the counter ourselves (if I understand it correctly from [libsodium](https://download.libsodium.org/doc/advanced/chacha20.html)). As for padding, do we need to create a block? Chacha20 is a stream cipher.

**Comment by wongsyrone** [comment]
Will https://github.com/jedisct1/libsodium.js help?

**Comment by nekolab** [comment]
@wongsyrone I don't think so, it seems doesn't contains chacha20 and salsa20 cipher support.
